raphtory/src/search/mod.rs

```diff
@@ -1,1498 +1,367 @@
+        // Create a new PropertyIndex if it doesn't exist
+        if prop_index_guard[prop_id].is_none() {
+            let mut schema_builder = PropertyIndex::schema_builder(&*prop_name, prop_type);
+            add_schema_fields(&mut schema_builder);
+            let schema = schema_builder.build();
+            let property_index = PropertyIndex::new(ArcStr::from(prop_name), schema);
+            let writer = property_index.index.writer(50_000_000)?;
 
-        let document = create_node_document(
-            node_id,
-            node_id_rev,
-            node_name,
-            node_type,
-            Box::new(constant_properties),
-            Box::new(temporal_properties),
-            &schema,
-        )?;
-        writer.add_document(document)?;
-        writer.commit()?;
-
-        self.graph.internal_add_node(t, v, props)
-    }
-
-    fn internal_add_edge(
-        &self,
-        t: TimeIndexEntry,
-        src: VID,
-        dst: VID,
-        props: &[(usize, Prop)],
-        layer: usize,
-    ) -> Result<MaybeNew<EID>, GraphError> {
-        let res = self.graph.internal_add_edge(t, src, dst, props, layer)?;
-
-        let edge_id = match res {
-            MaybeNew::New(idx) | MaybeNew::Existing(idx) => idx,
+            writers.push(Some(writer));
+            prop_index_guard[prop_id] = Some(property_index);
         }
-        .as_u64();
-        let src_name = self.graph.node_name(src);
-        let dst_name = self.graph.node_name(dst);
-
-        let constant_properties = iter::empty::<(ArcStr, Prop)>();
-        let temporal_properties: Box<dyn Iterator<Item = (i64, ArcStr, Prop)> + '_> =
-            Box::new(props.iter().filter_map(move |(prop_id, prop)| {
-                let prop_name = self.graph.edge_meta().get_prop_name(*prop_id, false);
-                Some((t.t(), prop_name, prop.clone()))
-            }));
-
-        let schema = self.edge_index.schema();
-        let mut writer = self.edge_index.writer(50_000_000)?;
-
-        let document = create_edge_document(
-            edge_id,
-            src_name,
-            dst_name,
-            Box::new(constant_properties),
-            Box::new(temporal_properties),
-            &schema,
-        )?;
-        writer.add_document(document)?;
-        writer.commit()?;
-
-        Ok(res)
     }
 
-    fn internal_add_edge_update(
-        &self,
-        t: TimeIndexEntry,
-        edge: EID,
-        props: &[(usize, Prop)],
-        layer: usize,
-    ) -> Result<(), GraphError> {
-        self.graph.internal_add_edge_update(t, edge, props, layer)
-    }
+    Ok(writers)
 }
 
-fn collect_constant_properties<'a>(
-    properties: impl Iterator<Item = (ArcStr, Prop)> + 'a,
-) -> serde_json::Map<String, serde_json::Value> {
-    properties.map(|(k, v)| (k.to_string(), v.into())).collect()
+fn initialize_node_const_property_indexes(
+    graph: &GraphStorage,
+    property_indexes: Arc<RwLock<Vec<Option<PropertyIndex>>>>,
```

Comment:
This needs to take a guard not the actual lock, just call `node_index.constant_property_indexes.write()`
you can set have  `property_indexes: impl DerefMut<Target = Vec<Option<PropertyIndex>>>` as a type here